presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -284,14 +281,21 @@ private Type getCommonSuperTypeForDecimal(DecimalType firstType, DecimalType sec
         return createDecimalType(targetPrecision, targetScale);
     }
 
-    private Type getCommonSuperTypeForVarchar(VarcharType firstType, VarcharType secondType)
+    private Type getCommonSuperTypeForCharacterStrings(Type firstType, Type secondType)
     {
-        return createVarcharType(Math.max(firstType.getLength(), secondType.getLength()));
+        int returnTypeLength = Math.max(getCharacterStringTypeLength(firstType), getCharacterStringTypeLength(secondType));
+        if (firstType instanceof VarcharType || secondType instanceof VarcharType) {
+            return createVarcharType(returnTypeLength);
+        }
+        return createCharType(returnTypeLength);
     }
 
-    private Type getCommonSuperTypeForChar(CharType firstType, CharType secondType)
+    private int getCharacterStringTypeLength(Type type)
     {
-        return createCharType(Math.max(firstType.getLength(), secondType.getLength()));
+        if (isVarcharType(type)) {
+            return ((VarcharType) type).getLength();
+        }
+        return checkType(type, CharType.class, "type").getLength();
```

Comment:
Replace check here with check at the beginning of the function testing if `type` is either `VarcharType` or `CharType`